File: ui/mocks/handlers/molq.ts
/**
 * Mock handlers for the Remote Operations (molq plugin) endpoints.
 */

import { http, HttpResponse } from "msw";

const API_BASE = "/api/plugins/molq";

const now = () => new Date().toISOString();

const targets = [
  {
    name: "dardel",
    scheduler: "slurm",
    clusterName: "dardel.scilifelab.se",
    jobsDir: "/scratch/jdoe/.molq/jobs",
    healthy: true,
    healthReason: null,
    activeJobs: 3,
  },
  {
    name: "alvis",
    scheduler: "slurm",
    clusterName: "alvis.scilifelab.se",
    jobsDir: null,
    healthy: true,
    healthReason: null,
    activeJobs: 1,
  },
  {
    name: "local",
    scheduler: "local",
    clusterName: "this-machine",
    jobsDir: null,
    healthy: true,
    healthReason: null,
    activeJobs: 0,
  },
  {
    name: "aws-gpu",
    scheduler: "slurm",
    clusterName: "aws.us-east-1",
    jobsDir: null,
    healthy: false,
    healthReason: "Connection refused",
    activeJobs: 0,
  },
];

const baseSubmitted = Date.now() - 4 * 3600 * 1000;

const buildJobs = () => {
  const states = ["running", "running", "queued", "succeeded", "succeeded", "failed", "running"];
  return states.map((state, idx) => {
    const submittedAt = new Date(baseSubmitted + idx * 5 * 60_000);
    const startedAt =
      state === "queued" ? null : new Date(submittedAt.getTime() + (60 + idx * 30) * 1000);
    const finishedAt =
      state === "succeeded" || state === "failed"
        ? new Date(submittedAt.getTime() + (3600 + idx * 60) * 1000)
        : null;
    const duration = startedAt
      ? ((finishedAt ?? new Date()).getTime() - startedAt.getTime()) / 1000
      : null;
    const target = idx % 2 === 0 ? "dardel" : "alvis";
    return {
      target,
      jobId: `job-${idx + 1}`,
      schedulerJobId: `${48290 + idx}`,
      clusterName: targets.find((t) => t.name === target)?.clusterName ?? null,
      scheduler: "slurm",
      name: ["train-allegro-lr-sweep", "nemd-conductivity-run", "gromacs-benchmark", "alphafold-batch-12", "ligand-docking-array", "md-ensemble-512", "data-prep-pipeline"][idx],
      state,
      submittedAt: submittedAt.toISOString(),
      startedAt: startedAt?.toISOString() ?? null,
      finishedAt: finishedAt?.toISOString() ?? null,
      exitCode: state === "succeeded" ? 0 : state === "failed" ? 1 : null,
      durationSeconds: duration,
      cwd: "/scratch/jdoe/runs",
    };
  });
};

const computeStats = (jobs: ReturnType<typeof buildJobs>) => {
  const running = jobs.filter((j) => j.state === "running").length;
  const pending = jobs.filter((j) => ["queued", "submitted", "created"].includes(j.state)).length;
  const failed = jobs.filter((j) =>
    ["failed", "timed_out", "cancelled", "lost"].includes(j.state),
  ).length;
  const succeeded = jobs.filter((j) => j.state === "succeeded").length;
  const waits = jobs
    .filter((j) => j.startedAt && j.submittedAt)
    .map(
      (j) => (new Date(j.startedAt as string).getTime() - new Date(j.submittedAt as string).getTime()) / 1000,
    );
  const avgWaitSeconds = waits.length
    ? waits.reduce((a, b) => a + b, 0) / waits.length
    : null;
  return { running, pending, failed, succeeded, avgWaitSeconds };
};

export const molqHandlers = [
  http.get(`${API_BASE}/targets`, () => {
    return HttpResponse.json({ targets, total: targets.length });
  }),

  http.get(`${API_BASE}/jobs`, ({ request }) => {
    const url = new URL(request.url);
    const target = url.searchParams.get("target");
    const all = buildJobs();
    const filtered = target ? all.filter((j) => j.target === target) : all;
    return HttpResponse.json({
      jobs: filtered,
      stats: computeStats(filtered),
      total: filtered.length,
    });
  }),

  http.get(`${API_BASE}/jobs/:jobId`, ({ params, request }) => {
    const url = new URL(request.url);
    const target = url.searchParams.get("target");
    const job = buildJobs().find(
      (j) => j.jobId === (params.jobId as string) && (!target || j.target === target),
    );
    if (!job) {
      return HttpResponse.json({ detail: "Job not found" }, { status: 404 });
    }
    return HttpResponse.json({
      ...job,
      failureReason: job.state === "failed" ? "Exit code 1" : null,
      metadata: { run_id: `run-${job.jobId}`, project: "demo" },
      commandDisplay: `python -m molexp.cli execute /scratch/${job.jobId}`,
      transitions: [
        {
          timestamp: job.submittedAt ?? now(),
          fromState: null,
          toState: "created",
          reason: "job created",
        },
        {
          timestamp: job.submittedAt ?? now(),
          fromState: "created",
          toState: "submitted",
          reason: null,
        },
        ...(job.startedAt
          ? [
              {
                timestamp: job.startedAt,
                fromState: "submitted",
                toState: "running",
                reason: null,
              },
            ]
          : []),
        ...(job.finishedAt
          ? [
              {
                timestamp: job.finishedAt,
                fromState: "running",
                toState: job.state,
                reason: null,
              },
            ]
          : []),
      ],
      upstreamTotal: 0,
      upstreamSatisfied: 0,
      downstreamTotal: 0,
    });
  }),

  http.get(`${API_BASE}/jobs/:jobId/logs`, ({ params }) => {
    const encoder = new TextEncoder();
    const lines = [
      `[INFO] Step 8200/10000 - loss: 0.2314 - lr: 2.1e-5`,
      `[INFO] Validation loss: 0.2167`,
      `[INFO] Checkpoint saved to /ckpts/step_8200.pt`,
      `[INFO] Step 8300/10000 - loss: 0.2281 - lr: 2.1e-5`,
      `[INFO] Validation loss: 0.2142`,
      `[INFO] (mock job ${params.jobId as string} stream end)`,
      "[stream closed]",
    ];
    const stream = new ReadableStream({
      start(controller) {
        let i = 0;
        const send = () => {
          if (i >= lines.length) {
            controller.close();
            return;
          }
          controller.enqueue(
            encoder.encode(`data: ${JSON.stringify({ line: lines[i] })}\n\n`),
          );
          i += 1;
          setTimeout(send, 200);
        };
        send();
      },
    });
    return new HttpResponse(stream, {
      headers: {
        "Content-Type": "text/event-stream",
        "Cache-Control": "no-cache",
      },
    });
  }),
];

File: src/molexp/server/schemas/molq.py
"""Response schemas for the molq Remote Operations endpoints."""

from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field

from molexp.plugins.submit_molq.dashboard import (
    JobDetail,
    JobsPage,
    JobSummary,
    QueueStats,
    TargetSummary,
)


class MolqTargetSummary(BaseModel):
    name: str
    scheduler: str
    clusterName: str | None = None
    jobsDir: str | None = None
    healthy: bool
    healthReason: str | None = None
    activeJobs: int

    @classmethod
    def from_dataclass(cls, target: TargetSummary) -> MolqTargetSummary:
        return cls(
            name=target.name,
            scheduler=target.scheduler,
            clusterName=target.cluster_name,
            jobsDir=target.jobs_dir,
            healthy=target.healthy,
            healthReason=target.health_reason,
            activeJobs=target.active_jobs,
        )


class MolqTargetListResponse(BaseModel):
    targets: list[MolqTargetSummary]
    total: int


class MolqJobSummary(BaseModel):
    target: str
    jobId: str
    schedulerJobId: str | None = None
    clusterName: str | None = None
    scheduler: str | None = None
    name: str | None = None
    state: str
    submittedAt: datetime | None = None
    startedAt: datetime | None = None
    finishedAt: datetime | None = None
    exitCode: int | None = None
    durationSeconds: float | None = None
    cwd: str | None = None

    @classmethod
    def from_dataclass(cls, job: JobSummary) -> MolqJobSummary:
        return cls(
            target=job.target,
            jobId=job.job_id,
            schedulerJobId=job.scheduler_job_id,
            clusterName=job.cluster_name,
            scheduler=job.scheduler,
            name=job.name,
            state=job.state,
            submittedAt=job.submitted_at,
            startedAt=job.started_at,
            finishedAt=job.finished_at,
            exitCode=job.exit_code,
            durationSeconds=job.duration_seconds,
            cwd=job.cwd,
        )


class MolqQueueStats(BaseModel):
    running: int = 0
    pending: int = 0
    failed: int = 0
    succeeded: int = 0
    avgWaitSeconds: float | None = None

    @classmethod
    def from_dataclass(cls, stats: QueueStats) -> MolqQueueStats:
        return cls(
            running=stats.running,
            pending=stats.pending,
            failed=stats.failed,
            succeeded=stats.succeeded,
            avgWaitSeconds=stats.avg_wait_seconds,
        )


class MolqJobsResponse(BaseModel):
    jobs: list[MolqJobSummary]
    stats: MolqQueueStats
    total: int

    @classmethod
    def from_page(cls, page: JobsPage) -> MolqJobsResponse:
        jobs = [MolqJobSummary.from_dataclass(j) for j in page.jobs]
        return cls(
            jobs=jobs,
            stats=MolqQueueStats.from_dataclass(page.stats),
            total=len(jobs),
        )


class MolqJobTransition(BaseModel):
    timestamp: datetime
    fromState: str | None = None
    toState: str
    reason: str | None = None


class MolqJobDetailResponse(MolqJobSummary):
    failureReason: str | None = None
    metadata: dict[str, str] = Field(default_factory=dict)
    commandDisplay: str | None = None
    transitions: list[MolqJobTransition] = Field(default_factory=list)
    upstreamTotal: int = 0
    upstreamSatisfied: int = 0
    downstreamTotal: int = 0

    @classmethod
    def from_dataclass(cls, detail: JobDetail) -> MolqJobDetailResponse:  # ty: ignore[invalid-method-override]
        base = MolqJobSummary.from_dataclass(detail.summary).model_dump()
        return cls(
            **base,
            failureReason=detail.failure_reason,
            metadata=detail.metadata,
            commandDisplay=detail.command_display,
            transitions=[
                MolqJobTransition(
                    timestamp=t.timestamp,
                    fromState=t.from_state,
                    toState=t.to_state,
                    reason=t.reason,
                )
                for t in detail.transitions
            ],
            upstreamTotal=detail.upstream_total,
            upstreamSatisfied=detail.upstream_satisfied,
            downstreamTotal=detail.downstream_total,
        )


__all__ = [
    "MolqJobDetailResponse",
    "MolqJobSummary",
    "MolqJobTransition",
    "MolqJobsResponse",
    "MolqQueueStats",
    "MolqTargetListResponse",
    "MolqTargetSummary",
]
